### DISCUSSION TITLE
Still works?

### DISCUSSION THREAD

**Discussion by Edward-Eth** [matheus0312/BrytonUtilities#6] [General]
Hi Matheus,

I've just discovered this github project from your post on Bikeforums.net from 2022, do you know if this code should still work? Looking at my Bryton 420's storage with it connected to my PC shows the following folders/files:
- PlanTrip folder (contains .FIT files with route names)
- Tracks folder (contains .SMY, .TINFO, .TRACK files for each route as well as a folder for each route containing a .PATH file and a .ZINFO file)

Additionally, did your project ever support translating waypoints from input routs into turn by turn directions the device can display?

Thanks,
Edward

**Comment by Edward-Eth** [comment]
Quick update from an evening of fiddling around with the code and features:
- Firstly, amazing work on the binary decoding of the .fit files, I wouldn't even know where to start on that.
- I've made a fork of the repo with my work in, so far I've:
  - Fixed a couple of minor bugs I saw
  - Added support for gpx files generated using plotaroute.com, my preferred source (this required a bit of fiddling, as it turns out that your code/bryton expects the instructions/waypoints to be interspersed in the route as it proceeds, but plotaroute puts them before the trackpoints as "wpt" items
  - minor QOL improvements to the code for readability

I still need to properly test my output .fit files, but loading them onto my rider shows a correct looking trace and shows instructions so it seems promising.

I don't know if you're still active on github/this project or still using this code, would be great to hear back if you are.

Edward

**Comment by matheus0312** [comment]
Hi Edward,

It is awesome to know that this project is helping someone else. 

When I started I working on it there was no documentation about Bryton's .fit file encoding (which to makes matter worse use the same extension as Garmin's), so I brute forced my way to understand what did what in it, basically created a couple of routes in Bryton app where I knew the total distance, elevation,... and looked through the .fit file to find this info.

What took me some time to discover, but in the end saved me a lot of time, is that only the .fit file is needed, since the other ones are generated by the unit when the .fit is decoded. Still it was good to understand how the other files were encoded to find my errors more easily.

For sure there are a lot of bugs to be fixed, it was one of my first (and only) big project in python. I implemented it initially almost in go horse mode using functional programming, and abandoned implementation when I started to modify it to apply some object oriented concepts (without success).

From what I remembered, I found no problems in routes that were correctly loaded in my unit. The only issue I found was a limit on the number of waypoints (or distance, my memory is bad), but these was shown when loading the unit. I found no new issues when cycling the routes.

Sadly I stopped using this code a year ago when I sold my Bryton. But if I can help somehow to decode my spaghetti code I will be happy to help :)

**Reply by Edward-Eth** [reply]
Hey Matheus, 

Thanks for getting back to me :) I'll let you know if I need any help figuring stuff out, generally (despite the lack of comments 😜 ) it's not too bad to understand, but maybe that's because I also write spaghetti code 😆 

One question I have is whether you ever looked at how the bryton .fit files handle roundabout instructions? Not sure how frequently you encounter those in your part of the world so maybe it was never relevant. It displays a number (nth exit) in a circle of arrows, but didn't see anything in your translation from gpx instructions to fit instructions that seemed likely to correspond to that.

Ed

**Reply by matheus0312** [reply]
Hi Edward,

Infere there are not a lot of roundabouts where I live, so I did not thought about decoding them 😕

**Comment by bagring** [comment]
Hi guys,
at first, Matheus, thank you for the code. I have also found it on the bikeforums.net. I have discovered that it actually **does** follow the common [Garmin .fit structure](https://developer.garmin.com/fit/protocol/). The only difference is that the Bryton uses some weird Global Message Numbers which are not defined in the FIT SDK.
I have collected the decoding info for Matheus but now I see that he has left the battlefield of the reverse engineering :-). Well, I will put it here for Edward... or anyone else. In particular, it reveals the mysterious "reserved" bytes in the `Decoding files.txt`.
[bryton_fit.txt](https://github.com/user-attachments/files/17008252/bryton_fit.txt)

Matheus, could you, please, put some info for the inexperienced users to let them know what to do with the resulting .fit file? The file(s) must be copied into the PlanTrip folder. Another surprise for me (as the first-time user) was that (probably) I cannot use any .gpx file. I have used the .gpx exported from [mapy.cz](https://mapy.cz/) and the resulting .fit did not contain turn instructions (apparently the .gpx has to contain the turn instructions already). Is the openrouteservice.org (and plotaroute.com thanks to Edward) the only service I can use to generate the .gpx (for the conversion to the .fit)? I prefer services which do not require any user account (offline app would be the best option as any website may be sooner or later taken out of service).

**Reply by Edward-Eth** [reply]
Hey Bagring,

The code can only extract existing directions from a GPX file that contains them, it doesn't seem like mapy.cz supports including turn directions in the gpx export, so unless someone were to write code to compare the gpx points to a world street map and back calculate turns on that map (sounds very complicated and time consuming), that site won't work.

At the moment the code is only able to extract openrouteservice and plotaroute gpx files. This is as GPX files are not totally consistent in formatting (why would they be, that would make life too easy 🙄). For example, openrouteservice includes the current direction on every trkpt, whereas plotaroute puts all instructions as waypoint objects at the start of the gpx, then just has trkpts with no directions afterwards. As such each different formatting of GPX files needs it's only translation into the standard list formatting the code wants routes in to export them.

I will add the instructions about importing the file into the device to my repo's readme, which is a bit more instructional than Matheus'.

I will also make a bit of a guide to making translators for different formats of gpx and include that in my readme, and you're welcome to add translators for any file type you'd like.

Edward

**Reply by bagring** [reply]
Hi Edward,
well, I did not dig into the code so I am just guessing.. ..but isn't it just comparison of two azimuths? Maybe there is some additional decision needed when the sections before and after the turn point are not straight. But it was just a note. I do not insist on using the .gpx from mapy.cz. I may register somewhere to get the .gpx with the turn data. It is not a big deal.

Well.. handling different .gpx formats requires some extra effort... but at least it is just a plain text, not some unknown binary blob where you have to guess, which bytes (or bits) belong together and what is their meaning. Btw I would parse the .gpx as a XML file rather than read line by line. I think it is cleaner and safer. See the attached example below.
[parse_gpx.zip](https://github.com/user-attachments/files/17021015/parse_gpx.zip)

Matheus, since you have sold your device, I guess you have no further motivation to maintain the repo. Please, could you note that in the README.md and forward the visitors to the Edward's one? I myself have noticed the Edward's fork just accidentally. Because of the bikeforums.net, the users will mostly end up here. I am afraid that when they realize that this repo is not maintained, they may start to create new repos or keep their modifications on their PCs. It would be nice to keep the community :relaxed: in one place.

Also... I would like to use your offer to help (if I can). As Edward is (i guess) interested more in the .gpx parsing part, I am (for now) interested rather in the understanding of the Bryton's .fit format. Do you remember, what is/could be [the "Alphabet"](https://github.com/matheus0312/BrytonUtilities/blob/master/doc/Decoding%20files.txt#L24-L28)? Was that always just a sequence of increasing numbers with length equal to number of instructions? It looks pretty useless.

Btw the generated .fit can be read with the Garmin SDK. It returns just single error related with the missing CRC at the end of the file. That is a good sign. Hats off to you, Matheus.
```python
from garmin_fit_sdk import Decoder, Stream

stream = Stream.from_file("SerraRioDoRastro.fit")
decoder = Decoder(stream)
messages, errors = decoder.read()
```

**Reply by Edward-Eth** [reply]
> Hi Edward, well, I did not dig into the code so I am just guessing.. ..but isn't it just comparison of two azimuths? Maybe there is some additional decision needed when the sections before and after the turn point are not straight. But it was just a note. I do not insist on using the .gpx from mapy.cz. I may register somewhere to get the .gpx with the turn data. It is not a big deal.

Hey, this could work for adding simply "left" "right" etc instructions to files without, but wouldn't allow us to put the street names that come with gpx instructions from other sites. And as you say there's limitations there based on what amount of turn constitutes a turn worth flagging vs just a bend in a road.

**Reply by Edward-Eth** [reply]
@bagring could you do me a favour and see if you're able to make a branch on my repo, commit something and raise a merge request? I'm not sure if I have the repo set up right to allow people to contribute but keen for that to be the case.

**Reply by matheus0312** [reply]
Hey guys,

If it helps you guys too, I used gpx files from openstreetmap which is free and public.

I will update the readme to point to Edward's project

**Comment by bagring** [comment]
I tried but I was not successful. But maybe it is just me. I have almost no experience with GitHub. I was using this account just to ask someone to fix something so far :-).
I have installed git-scm & TortoiseGit. I have cloned your repo, made a change, commit to different branch, but push returns "fatal: Authentication failed for 'https://github.com/Edward-Eth/BrytonUtilities24.git/'". It offers this dialog:
![obrazek](https://github.com/user-attachments/assets/57e36bd1-05f5-47db-827f-e62a0de88d97)
Clicking first option ends up with:
![obrazek](https://github.com/user-attachments/assets/9ff2c8f8-7455-489d-8208-1ca72eb1d6ad)
When i choose the second option, it ends up with:
![obrazek](https://github.com/user-attachments/assets/7d951f8f-0d49-41d8-b4eb-0d3728c4fa37)

...but the final result is the same in both the cases:
![obrazek](https://github.com/user-attachments/assets/9edda495-df30-4e4c-90ea-06c2f7c366c8)

**Reply by bagring** [reply]
I would like to ask for the explanation of [these lines](https://github.com/Edward-Eth/BrytonUtilities24/blob/44adffe089813a0d802a2ec0e0be7a050f7a68f9/BrytonUtilities/gpx_utilities.py#L61C1-L63C36). You are omitting consecutive instructions with same name. Is that safe? I mean what if there are just just unnamed roads and you are about to turn left, right,... or even making a u-turn (left, left)? What if it is a valid scenario?

**Reply by Edward-Eth** [reply]
> I would like to ask for the explanation of [these lines](https://github.com/Edward-Eth/BrytonUtilities24/blob/44adffe089813a0d802a2ec0e0be7a050f7a68f9/BrytonUtilities/gpx_utilities.py#L61C1-L63C36). You are omitting consecutive instructions with same name. Is that safe? I mean what if there are just just unnamed roads and you are about to turn left, right,... or even making a u-turn (left, left)? What if it is a valid scenario?

Hey Bagring, 

I've invited you as a collaborator on the github project, that might make it work now?

As for the instructions thing, that's code from Mattheus originally but the reasoning is this:
- In open route service GPX files, every single point on the path contains an instruction, eg:
![image](https://github.com/user-attachments/assets/1b54f8df-49b7-4e54-b1f0-972562493d00)
- If we put every one of these as an instruction into the .fit file for the 420, it would constantly say "xyz instruction in 0m" and beep on every waypoint
- We only therefore want the last one of these instructions to be input, as this is the last one before the turn actually happens

The only time this would be an issue is if the name (in the screenshot "Rodovia Governador Irineu Bornhausen, SC-390") was the same for two actual instructions that are actually different things. As you say, this could occur if you had "unnamed road" twice in a row, so it could be worth changing it to check if the type is also the same, even then it would not help if it was a left turn followed by a left turn.

If you have any ideas on how to avoid skipping two instructions of the same type both on the same road name that would be helpful, but we don't want to remove this code as it would break instructions on routes from this source.

Routes from Plotaroute.com don't have this issue, as they only output instructions on the point where the turn occurs.

Also, probably best to raise any future questions/problems as issues or discussions on my repo instead of on Matheus's repo as he's no longer involved.

Edward

**Reply by bagring** [reply]
Thanks, I will try it once I will have some spare time.

As for the instructions, I was not asking for removing the code. I just wanted to ask whether you are aware of that. At the moment I do not see any way how to recognize whether there are multiple same (and all legit) instructions or just repetitions of the single one. I think this should be fixed in the ORS. Maybe there could be some flag/switch/attribute to change this behavior? I know you are focusing on the plotaroute.com, so please, take it just as a side note.
Btw. I think we are taking the _first_ one of these instructions (not the last one).

I agree to move the discussion on your repo so that we are not bothering Matheus anymore.

Matheus, once again thank you for all the effort you put into this.
